source/common/common/utility.h

```diff
@@ -636,15 +636,100 @@ class WelfordStandardDeviation {
   double m2_{0};
 };
 
-template <class Value> struct TrieEntry {
+// FastTrieEntry is used for very simple very fast lookups.
+//
+// Each node contains 256 pointers to potential child nodes, making an
+// 8KB allocation for every node. Since the vast majority of nodes are
+// very sparse, this is extremely space-inefficient, but it makes for
+// the fastest possible lookups if the total table size is small.
+// At larger table sizes, memory cache misses come into play and a more
+// compact trie actually runs faster.
+//
+// Benchmarks suggest that for the important use case of
+// the static header lookup table, FastTrieEntry lookups run in about
+// 2/3 the time that SmallTrieEntry lookups do, but for cases like
+// large route tables, or with long keys, FastTrieEntry runs in around
+// 2x the time and 100x the memory.
+//
+// FastTrieEntry should therefore should not be used in contexts where
+// performance is not critical, or where the contents may be large.
+template <class Value> class FastTrieEntry {
+public:
+  FastTrieEntry* operator[](uint8_t i) const { return entries_[i].get(); }
+  void set(uint8_t branch, std::unique_ptr<FastTrieEntry<Value>> entry) {
+    entries_[branch] = std::move(entry);
+  }
   Value value_{};
-  std::array<std::unique_ptr<TrieEntry>, 256> entries_;
+
+private:
+  std::array<std::unique_ptr<FastTrieEntry>, 256> entries_;
+};
+
+// A SmallTrieEntry aims to be a good balance of performant and
+// space-efficient, by allocating a vector the size of the range of children
+// the node contains. This should be good for most use-cases.
+//
+// For example, a node with children 'a' and 'z' will contain a vector of
+// size 26, containing two values and 24 nulls. A node with only one
```

Comment:
did you try using a tree rather than a vector? Something like absl::btree_set ?

I'd be interested to see if that might give good performance in all the cases.

Reply:
Lookups through a btree would be multiple steps. O(log(num_keys)) is surely worse than O(1) when the 1 is also very simple.